libs/telemetry/logger.go

```diff
@@ -0,0 +1,139 @@
+package telemetry
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+	"time"
+
+	"github.com/databricks/cli/libs/log"
+	"github.com/google/uuid"
+)
+
+// Interface abstraction created to mock out the Databricks client for testing.
+type DatabricksApiClient interface {
+	Do(ctx context.Context, method, path string,
+		headers map[string]string, request, response any,
+		visitors ...func(*http.Request) error) error
+}
+
+type Logger interface {
+	// Record a telemetry event, to be flushed later.
+	Log(event DatabricksCliLog)
+
+	// Flush all the telemetry events that have been logged so far. We expect
+	// this to be called once per CLI command for the default logger.
+	Flush(ctx context.Context, apiClient DatabricksApiClient)
+
+	// This function is meant to be only to be used in tests to introspect
+	// the telemetry logs that have been logged so far.
+	Introspect() []DatabricksCliLog
+}
+
+type defaultLogger struct {
+	logs []FrontendLog
+}
+
+func (l *defaultLogger) Log(event DatabricksCliLog) {
+	if l.logs == nil {
+		l.logs = make([]FrontendLog, 0)
+	}
+	l.logs = append(l.logs, FrontendLog{
+		// The telemetry endpoint deduplicates logs based on the FrontendLogEventID.
+		// This it's important to generate a unique ID for each log event.
+		FrontendLogEventID: uuid.New().String(),
+		Entry: FrontendLogEntry{
+			DatabricksCliLog: event,
+		},
+	})
+}
+
+// Maximum additional time to wait for the telemetry event to flush. We expect the flush
+// method to be called when the CLI command is about to exist, so this caps the maximum
+// additional time the user will experience because of us logging CLI telemetry.
+var MaxAdditionalWaitTime = 3 * time.Second
+
+// We make the API call to the /telemetry-ext endpoint to log the CLI telemetry events
+// right about as the CLI command is about to exit. The API endpoint can handle
+// payloads with ~1000 events easily. Thus we log all the events at once instead of
+// batching the logs across multiple API calls.
+func (l *defaultLogger) Flush(ctx context.Context, apiClient DatabricksApiClient) {
```

Comment:
In the future, this function could spawn a new process to avoid the latency cost altogether. Flushing the events to disk and having a new process consume them seems more reliable than this timeout based approach.